### PR TITLE
[SPARK-48318][SQL] Enable hash join support for all collations (complex types)

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Enable collation support for hash join on complex types.

- Logical plan is rewritten in analysis to (recursively) replace all non-binary strings with CollationKey
- CollationKey is a unary expression that transforms StringType to BinaryType
- Collation keys allow correct & efficient string comparison under specific collation rules

### Why are the changes needed?
Improve JOIN performance for complex types containing collated strings.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

- Unit tests for `CollationKey` in `CollationExpressionSuite`
- E2e SQL tests for `RewriteCollationJoin` in `CollationSuite`
- Various queries with JOIN in existing TPCDS collation test suite

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala [hunk 1]
```diff
@@ -17,29 +17,68 @@
...
+      case st: StringType
+        if !CollationFactory.fetchCollation(st.collationId).supportsBinaryEquality =>
+          CollationKey(expr)
+
+      case StructType(fields) =>
+        processStruct(expr, fields)
+
+      case ArrayType(et, containsNull) =>
+        processArray(expr, et, containsNull)
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:53]: how about map type? I think we support map type join/grouping keys now?

Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:53]: I don't think joins and aggregations are supported for map type at this time

join:
```
Failed to analyze query: org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.INVALID_ORDERING_TYPE] Cannot resolve "(m = m)" due to data type mismatch: The `=` does not support ordering on type "MAP<STRING, STRING>"
```

agg:
```
Failed to analyze query: org.apache.spark.sql.AnalysisException: [UNSUPPORTED_FEATURE.SET_OPERATION_ON_MAP_TYPE] The feature is not supported: Cannot have MAP type columns in DataFrame which calls set operations (INTERSECT, EXCEPT, etc.), but the type of column `m` is "MAP<STRING, STRING>".
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 2]
```diff
@@ -1046,6 +1046,211 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+
+    case class HashJoinTestCase[R](collation: String, result: R)
+    val testCases = Seq(
+      HashJoinTestCase("UTF8_BINARY",
+        Seq(Row(Seq("aa"), 1, Seq("aa"), 2))),
+      HashJoinTestCase("UTF8_BINARY_LCASE",
+        Seq(Row(Seq("aa"), 1, Seq("AA"), 2), Row(Seq("aa"), 1, Seq("aa"), 2))),
+      HashJoinTestCase("UNICODE",
+        Seq(Row(Seq("aa"), 1, Seq("aa"), 2))),
+      HashJoinTestCase("UNICODE_CI",
```
Comment (dbatomic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1061]: Can we start adding at least one locale + case insensitive + accent insensitive variation in tests? E.g. I would like to see if things work correctly for SRB_CI_AI, SRB_CS_AS, SRB_CI_AS and SRB_CS_AI.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala [hunk 3]
```diff
@@ -17,29 +17,68 @@
...
+          e.withNewChildren(Seq(processExpression(l, l.dataType), processExpression(r, r.dataType)))
       }
       if (!newCondition.fastEquals(condition)) {
         j.copy(condition = Some(newCondition))
       } else {
         j
       }
   }
+
+  private def processExpression(expr: Expression, dt: DataType): Expression = {
```
Comment (dbatomic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:42]: Hey, can you add a comment explaining what is happening here.
I think that it is not obvious why we are inserting `CollationKey` as a rule. I would appreciate comment header on top of `RewriteCollationJoin` that explains why and what we are doing here.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 4]
```diff
@@ -1046,6 +1046,211 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+          }.nonEmpty
+        )
+        assert(
+          collectFirst(queryPlan) {
+            case _: ShuffledJoin => ()
+          }.isEmpty
+        )
+
+        // if collation doesn't support binary equality, collation key should be injected
+        if (!CollationFactory.fetchCollation(t.collation).supportsBinaryEquality) {
```
Comment (dbatomic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1091]: Can you:
1) Verify that there is no array transform if collation is not binary.
2) Check that there is CollationKey as part of ArrayTransform function.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala [hunk 5]
```diff
@@ -17,29 +17,68 @@
...
+      case ArrayType(et, containsNull) =>
+        processArray(expr, et, containsNull)
+
+      case _ =>
+        expr
+    }
+  }
+
+  private def processStruct(str: Expression, fields: Array[StructField]): Expression = {
+    val struct = CreateNamedStruct(fields.zipWithIndex.flatMap {
```
Comment (dbatomic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:60]: do you need case here? Why not just flatMap((f,i) => { })?

**Review by dbatomic — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala [hunk 6]
```diff
@@ -17,29 +17,68 @@
...
+      case (f, i) =>
+        Seq(Literal(f.name),
+          processExpression(GetStructField(str, i, Some(f.name)), f.dataType))
+    }.toImmutableArraySeq)
+    if (struct.valExprs.forall(_.isInstanceOf[GetStructField])) {
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:65]: Or even better, we can short-circuit right away for all binary stable expressions in `processExpression`.

As for the naming, all these methods are private and internal to `RewriteCollationJoin` - I've added comments explaining what's going on, so I hope this should be enough

**Comment by uros-db** [comment]
@cloud-fan ready for further review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala [hunk 7]
```diff
@@ -17,29 +17,83 @@
...
+      // Joins are not supported on maps, so there's no special handling for MapType.
+      case _ =>
+        expr
+    }
+  }
+
+  private def processStruct(str: Expression, fields: Array[StructField]): Expression = {
+    val struct = CreateNamedStruct(fields.zipWithIndex.flatMap { case (f, i) =>
+      Seq(Literal(f.name), processExpression(GetStructField(str, i, Some(f.name)), f.dataType))
+    }.toImmutableArraySeq)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:81]: shall we return the original `str` expression if no field is updated with CollationKey?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala (hunk 7 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala:81]: `processExpression` should handle that by short-circuiting in the first `case`

furthermore, if all fields in a struct will remain unchanged - `processStruct` won't even get called to begin with

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**